lib/rdoc/rdoc.rb

```diff
@@ -449,6 +455,9 @@ def document options
     end
     @options.finish
 
+    ::RDoc::BasePlugin.activate_with(self)
+    @options.load_plugins
```

Comment:
If we load plugins dynamically and use `require`, plugins will not be loaded to multiple RDoc instance.

Do we need to load plugins for each `RDoc::RDoc#document`?

Reply:
Agree, `initialize` is a better place?

Reply:
`#initialize` is better than `#document` but do we need to load plugins for each RDoc instance?
Can we load plugins only once in one process?

Reply:
I cannot see that using multiple `RDoc::RDoc` instance is common. From CLI or Rake task, I think we have only one instance of `RDoc::RDoc`.
However, we might have multiple instances in the future, probably for concurrency.
So then, where is the ideal place to activate plugins? I through it's `RDoc::Rdoc` since it's in the execution path anyway.

Reply:
`gem install A B C` creates a `RDoc::RDoc` for each package.

Reply:
I see, that's pretty common...
Would we want separate plugins and their configurations for each `RDoc::RDoc` instance? That said, if we want to isolate each environment, loading plugins for each instance is fine. If not, then we can load plugins all at once at somewhere else (with `RDoc` module maybe?)

Reply:
Who does configure RDoc plugins?

If each package have RDoc plugins configuration, we may need per package plugins configuration. If users configure RDoc plugins, we may be able to share the same plugins configuration.

Reply:
I thought only users will configure RDoc plugins, but noticed that each gem can configure RDoc along with its plugins with `rdoc_options`.

https://github.com/ruby/rdoc/blob/9f0b3adda279dd5a5c2229b65204c11b37056232/lib/rdoc/rubygems_hook.rb#L167

And I guess that's why we instantiate an RDoc instance for each package in the first place. So loading plugins for each package might be necessary.

Reply:
We may not need to `require` for registering a plugin. We may just need to instantiate a plugin for it. For example:

```ruby
module RDoc
  class YardPlugin < BasePlugin
    def on_rdoc_store_complete(store)
      # ...
    end
  end

  class RDoc
    def document(options)
      # ...
      plugins = options.plugins.collect do |name|
        instantiate_plugin(name) # name.capitalize.constantize.new or associating name with class or something...
      end
      # ...
      plugins.each do |plugin|
        plugin.on_rdoc_store_complete(@store) if plugin.respond_to?(:on_rdoc_store_complete)
      end
      # ...
    end
  end
end
```

Reply:
Yes this is absolutely possible, in this case `on_some_event` style looks great and seems better than `listens` DSL style.